File: frontend/js/pages/analytics.js
/* Analytics: consumption, reorder, expiry risk, forecast, turnover,
   purchasing patterns. Tabs share one page; data loads per tab. */

import { api, badge, daysLabel, formatDate, html, money, mount, number, pageHeader, sortableTable, statTile } from "../core.js";

const TABS = [
    ["reorder", "Reorder"],
    ["risk", "Expiry risk"],
    ["consumption", "Consumption"],
    ["forecast", "Forecast"],
    ["turnover", "Turnover"],
    ["valuation", "Valuation"],
    ["purchasing", "Purchasing"],
];

/* Single-series inline bar: the number is always printed beside it. */
function bar(value, max) {
    const pct = max > 0 && value > 0 ? Math.max(2, Math.round((value / max) * 100)) : 0;
    return html`<span class="bar" aria-hidden="true"><span class="bar-fill" data-pct="${pct}"></span></span>`;
}

function applyBars(root) {
    root.querySelectorAll(".bar-fill[data-pct]").forEach(el => { el.style.width = `${el.dataset.pct}%`; });
}

const RENDERERS = {
    async reorder(el) {
        const rows = await api("/analytics/reorder");
        const needed = rows.filter(r => r.reorder_recommended);
        mount(el, html`
            <p class="method">A reorder is recommended when usable stock is at or below the reorder level, or will run out within the supplier lead time. The suggested quantity covers lead time + cover days at the current consumption rate, minus stock already on order.</p>
            <div class="cards compact">${statTile("Reorders recommended", number(needed.length), "orange")}
                ${statTile("Units to order", number(needed.reduce((s, r) => s + r.recommended_quantity, 0)), "blue")}</div>
            <div id="t"></div>`);
        sortableTable(el.querySelector("#t"), "reorder-table", [
            { label: "Medicine", key: "medicine", render: r => html`<a href="#/medicines/${r.medicine_id}"><strong>${r.medicine}</strong></a> <small>${r.strength}</small>` },
            { label: "Usable", key: "usable_stock", render: r => number(r.usable_stock), className: "num" },
            { label: "Reorder lvl", key: "reorder_level", className: "num" },
            { label: "Daily use", key: "average_daily_consumption", render: r => number(r.average_daily_consumption, 2), className: "num" },
            { label: "Days of stock", key: "days_of_stock", render: r => r.days_of_stock ?? "—", className: "num" },
            { label: "On order", key: "on_order", className: "num" },
            { label: "Suggested", key: "recommended_quantity", render: r => r.reorder_recommended ? html`<strong>${number(r.recommended_quantity)}</strong>` : "—", className: "num" },
            { label: "Status", key: "stock_status", render: r => badge(r.stock_status) },
            { label: "Why", key: "reason", render: r => r.reason || "—" },
        ], rows);
    },

    async risk(el) {
        const data = await api("/analytics/expiry-risk");
        mount(el, html`
            <p class="method">Batches are consumed earliest-expiry-first at each medicine's current daily consumption. Units still on the shelf at a batch's expiry date are "at risk". HIGH: units at risk inside the urgent window, or at least half the batch at risk inside the approaching window. MEDIUM: at risk further out (likely dead stock unless demand changes).</p>
            <div class="cards compact">
                ${statTile("Batches at risk", number(data.summary.batches_at_risk), "red")}
                ${statTile("Units at risk", number(data.summary.units_at_risk), "orange")}
                ${statTile("Value at risk", money(data.summary.value_at_risk), "red",
                    data.summary.batches_without_cost ? `${data.summary.batches_without_cost} batches without cost` : "")}
            </div>
            <div id="t"></div>`);
        const max = Math.max(0, ...data.batches.map(r => r.projected_units_at_risk));
        sortableTable(el.querySelector("#t"), "risk-table", [
            { label: "Risk", key: "risk_level", render: r => badge(r.risk_level), sort: r => ["EXPIRED", "HIGH", "MEDIUM", "LOW"].indexOf(r.risk_level) },
            { label: "Medicine", key: "medicine", render: r => html`<a href="#/medicines/${r.medicine_id}">${r.medicine}</a> <small>${r.strength}</small>` },
            { label: "Batch", key: "batch_number", render: r => html`<a href="#/batches/${r.batch_id}">${r.batch_number}</a>` },
            { label: "Expiry", key: "expiry_date", render: r => `${formatDate(r.expiry_date)} (${daysLabel(r.days_until_expiry)})` },
            { label: "Qty", key: "quantity", className: "num" },
            { label: "Daily use", key: "average_daily_consumption", render: r => number(r.average_daily_consumption, 2), className: "num" },
            { label: "Units at risk", key: "projected_units_at_risk", render: r => html`${bar(r.projected_units_at_risk, max)} ${number(r.projected_units_at_risk)}` },
            { label: "Value at risk", key: "value_at_risk", render: r => money(r.value_at_risk), className: "num" },
        ], data.batches, { afterRender: applyBars });
    },

    async consumption(el) {
        const data = await api("/analytics/consumption");
        mount(el, html`
            <p class="method">Units dispensed. Daily average uses the last 30 days (or 90 days when nothing was dispensed recently). Trend compares the last 30 days with the 30 days before.</p>
            <div id="t"></div>`);
        const max = Math.max(0, ...data.medicines.map(r => r.units_dispensed_last_90_days));
        sortableTable(el.querySelector("#t"), "consumption-table", [
            { label: "Medicine", key: "medicine", render: r => html`<a href="#/medicines/${r.medicine_id}">${r.medicine}</a> <small>${r.strength}</small>` },
            { label: "30 days", key: "units_dispensed_last_30_days", className: "num" },
            { label: "90 days", key: "units_dispensed_last_90_days", render: r => html`${bar(r.units_dispensed_last_90_days, max)} ${number(r.units_dispensed_last_90_days)}` },
            { label: "Daily avg", key: "average_daily_consumption", render: r => number(r.average_daily_consumption, 2), className: "num" },
            { label: "Weekly avg", key: "average_weekly_consumption", render: r => number(r.average_weekly_consumption, 1), className: "num" },
            { label: "Trend", key: "trend", render: r => html`${badge(r.trend)} ${r.trend_change_percent !== null ? `${r.trend_change_percent > 0 ? "+" : ""}${r.trend_change_percent}%` : ""}` },
            { label: "Days of stock", key: "days_of_stock", render: r => r.days_of_stock ?? "—", className: "num" },
            { label: "Class", key: "movement_class", render: r => badge(r.movement_class) },
        ], data.medicines.sort((a, b) => b.units_dispensed_last_90_days - a.units_dispensed_last_90_days), { afterRender: applyBars });
    },

    async forecast(el) {
        const rows = await api("/analytics/forecast", { params: { horizon_days: 30 } });
        mount(el, html`
            <p class="method">${rows[0]?.method || ""}. Confidence reflects how many of the last 12 weeks had dispensing activity. Forecasts support, not replace, purchasing judgement.</p>
            <div id="t"></div>`);
        sortableTable(el.querySelector("#t"), "forecast-table", [
            { label: "Medicine", key: "medicine", render: r => html`<a href="#/medicines/${r.medicine_id}">${r.medicine}</a> <small>${r.strength}</small>` },
            { label: "Weekly demand", key: "forecast_weekly_demand", render: r => number(r.forecast_weekly_demand, 1), className: "num" },
            { label: "30-day demand", key: "forecast_demand", render: r => number(r.forecast_demand, 1), className: "num" },
            { label: "Usable stock", key: "usable_stock", className: "num" },
            { label: "Stock after 30 d", key: "projected_stock_at_horizon", render: r => number(r.projected_stock_at_horizon, 1), className: "num" },
            { label: "Covered", key: "covers_horizon", render: r => r.covers_horizon ? badge("Covered", "NORMAL") : badge("Shortfall", "EXPIRED") },
            { label: "Confidence", key: "confidence", render: r => badge(r.confidence, r.confidence === "HIGH" ? "NORMAL" : r.confidence === "MEDIUM" ? "URGENT" : "NO MOVEMENT") },
        ], rows);
    },

    async turnover(el) {
        const rows = await api("/analytics/turnover");
        mount(el, html`
            <p class="method">Turnover = units dispensed in 90 days ÷ average stock (mean of today's stock and stock 90 days ago, reconstructed from the movement ledger).</p>
            <div id="t"></div>`);
        sortableTable(el.querySelector("#t"), "turnover-table", [
            { label: "Medicine", key: "medicine", render: r => html`${r.medicine} <small>${r.strength}</small>` },
            { label: "Current stock", key: "current_stock", className: "num" },
            { label: "Avg stock", key: "average_stock", className: "num" },
            { label: "Dispensed 90 d", key: "dispensed_90d", className: "num" },
            { label: "Turnover (90 d)", key: "turnover_90d", render: r => r.turnover_90d ?? "—", className: "num" },
            { label: "Annualised", key: "annualised_turnover", render: r => r.annualised_turnover ?? "—", className: "num" },
            { label: "Days of inventory", key: "days_of_inventory", render: r => r.days_of_inventory ?? "—", className: "num" },
        ], rows);
    },

    async valuation(el) {
        const v = await api("/analytics/valuation");
        const statuses = Object.entries(v.by_expiry_status);
        mount(el, html`
            <p class="method">Stock valued at batch acquisition cost. Units with no recorded cost are counted but not valued (never estimated).</p>
            <div class="cards compact">
                ${statTile("Total value", money(v.total.value), "blue", `${number(v.total.units)} units`)}
                ${statTile("Usable value", money(v.usable_value), "green")}
                ${statTile("Expired value", money(v.expired.value), "red")}
                ${statTile("Units without cost", number(v.total.units_without_cost), "orange")}
            </div>
            <div class="lower-grid">
                <div id="t1"></div><div id="t2"></div>
            </div>`);
        sortableTable(el.querySelector("#t1"), "val-status", [
            { label: "Expiry status", render: ([s]) => badge(s) },
            { label: "Batches", render: ([, x]) => number(x.batches), className: "num" },
            { label: "Units", render: ([, x]) => number(x.units), className: "num" },
            { label: "Value", render: ([, x]) => money(x.value), className: "num" },
        ], statuses);
        sortableTable(el.querySelector("#t2"), "val-location", [
            { label: "Location", key: "location" },
            { label: "Batches", key: "batches", className: "num" },
            { label: "Units", key: "units", render: r => number(r.units), className: "num" },
            { label: "Value", key: "value", render: r => money(r.value), className: "num" },
        ], v.by_location);
    },

    async purchasing(el) {
        const p = await api("/analytics/purchasing");
        mount(el, html`<div class="lower-grid"><div id="t1"></div><div id="t2"></div></div>`);
        sortableTable(el.querySelector("#t1"), "pur-supplier", [
            { label: "Supplier", key: "supplier" },
            { label: "Orders", key: "orders", className: "num" },
            { label: "Ordered value", key: "ordered_value", render: r => money(r.ordered_value), className: "num" },
            { label: "Fill rate", key: "fill_rate_percent", render: r => r.fill_rate_percent === null ? "—" : `${r.fill_rate_percent}%`, className: "num" },
            { label: "Avg lead time", key: "average_lead_time_days", render: r => r.average_lead_time_days === null ? "—" : `${r.average_lead_time_days} d`, className: "num" },
        ], p.by_supplier);
        sortableTable(el.querySelector("#t2"), "pur-month", [
            { label: "Month", key: "month" },
            { label: "Orders", key: "orders", className: "num" },
            { label: "Value", key: "ordered_value", render: r => money(r.ordered_value), className: "num" },
        ], p.by_month, { empty: "No orders." });
    },
};

export async function render(ctx) {
    const active = TABS.some(([key]) => key === ctx.params.query?.tab) ? ctx.params.query.tab : "reorder";
    mount(ctx.main, html`
        ${pageHeader("Analytics", "Consumption, reorder, expiry risk, forecasting and turnover — computed from live data")}
        <nav class="tabs" aria-label="Analytics views">
            ${TABS.map(([key, label]) => html`<a href="#/analytics?tab=${key}" class="tab ${key === active ? "active" : ""}"
                ${key === active ? html`aria-current="page"` : ""}>${label}</a>`)}
        </nav>
        <section class="section" id="tab-body"><div class="loading">Loading…</div></section>`);
    await RENDERERS[active](ctx.main.querySelector("#tab-body"));
}

File: backend/services/analytics.py
# ============================================================
# INVENTORY INTELLIGENCE
# ============================================================
# Low stock and reorder recommendations, consumption, slow movers, expiry
# risk, valuation, forecasting, turnover and purchasing patterns. All
# figures are computed from the live database; every method is described in
# the returned data so users can see how a number was produced.

import math
from datetime import date

import psycopg

from . import app_settings, expiry, inventory


def _settings(conn):
    values = app_settings.get_all(conn)
    return {
        "slow_units": int(values["stock.slow_moving_units_90d"]),
        "lead_time": int(values["reorder.lead_time_days"]),
        "cover_days": int(values["reorder.cover_days"]),
    }


# ------------------------------------------------------------
# Low stock and reorder recommendations
# ------------------------------------------------------------

def reorder_recommendations(conn: psycopg.Connection, *, only_needed: bool = False) -> list[dict]:
    """Per medicine: usable stock, consumption rate, days of stock remaining
    and a reorder recommendation.

    A reorder is recommended when usable stock is at/below the reorder level
    or will run out within the supplier lead time. The quantity brings stock
    (including units already on order) up to
        max(reorder_level, daily_rate * (lead_time + cover_days)).
    """
    cfg = _settings(conn)
    stock = inventory.medicine_stock(conn)
    usage = inventory.consumption(conn)
    on_order = inventory.outstanding_orders(conn)

    results = []
    for item in stock:
        use = usage.get(item["medicine_id"], {})
        daily = use.get("average_daily", 0) or 0
        usable = item["usable_stock"]
        ordered = on_order.get(item["medicine_id"], 0)
        days_of_stock = round(usable / daily, 1) if daily > 0 else None

        below_level = usable <= item["reorder_level"]
        runs_out_in_lead_time = days_of_stock is not None and days_of_stock <= cfg["lead_time"]
        target = max(item["reorder_level"], math.ceil(daily * (cfg["lead_time"] + cfg["cover_days"])))
        shortfall = max(0, target - usable - ordered)

        needed = (below_level or runs_out_in_lead_time) and shortfall > 0
        if needed:
            reasons = []
            if usable == 0:
                reasons.append("out of usable stock")
            elif below_level:
                reasons.append(f"usable stock {usable} is at or below reorder level {item['reorder_level']}")
            if runs_out_in_lead_time:
                reasons.append(f"about {days_of_stock} days of stock left, lead time is {cfg['lead_time']} days")
            reason = "; ".join(reasons)
        elif (below_level or runs_out_in_lead_time) and ordered:
            reason = f"low, but {ordered} units already on order"
        else:
            reason = None

        row = {
            "medicine_id": item["medicine_id"],
            "medicine": item["medicine"],
            "strength": item["strength"],
            "dosage_form": item["dosage_form"],
            "usable_stock": usable,
            "expired_stock": item["expired_stock"],
            "reorder_level": item["reorder_level"],
            "stock_status": item["stock_status"],
            "on_order": ordered,
            "average_daily_consumption": daily,
            "days_of_stock": days_of_stock,
            "target_stock": target,
            "reorder_recommended": needed,
            "recommended_quantity": shortfall if needed else 0,
            "reason": reason,
        }
        if not only_needed or needed:
            results.append(row)

    results.sort(key=lambda r: (not r["reorder_recommended"], r["days_of_stock"] if r["days_of_stock"] is not None else 1e9, r["medicine"]))
    return results


# ------------------------------------------------------------
# Consumption, slow and fast movers
# ------------------------------------------------------------

def consumption_summary(conn: psycopg.Connection) -> list[dict]:
    stock = {s["medicine_id"]: s for s in inventory.medicine_stock(conn)}
    usage = inventory.consumption(conn)
    cfg = _settings(conn)

    rows = []
    for medicine_id, use in usage.items():
        item = stock[medicine_id]
        d90 = use["dispensed_90d"]
        if d90 == 0:
            movement_class = "NO MOVEMENT"
        elif d90 <= cfg["slow_units"]:
            movement_class = "SLOW-MOVING"
        else:
            movement_class = "ACTIVE"
        daily = use["average_daily"]
        rows.append({
            "medicine_id": medicine_id,
            "medicine": item["medicine"],
            "strength": item["strength"],
            "dosage_form": item["dosage_form"],
            "units_dispensed_last_30_days": use["dispensed_30d"],
            "units_dispensed_previous_30_days": use["dispensed_prev_30d"],
            "units_dispensed_last_90_days": d90,
            "dispensing_transactions_last_90_days": use["transactions_90d"],
            "average_daily_consumption": daily,
            "average_weekly_consumption": use["average_weekly"],
            "trend": use["trend"],
            "trend_change_percent": use["trend_change_percent"],
            "last_dispensed": use["last_dispensed"],
            "usable_stock": item["usable_stock"],
            "days_of_stock": round(item["usable_stock"] / daily, 1) if daily > 0 else None,
            "movement_class": movement_class,
        })
    return rows


# ------------------------------------------------------------
# Expiry risk
# ------------------------------------------------------------

def expiry_risk(conn: psycopg.Connection, *, include_normal: bool = True) -> list[dict]:
    """Batch-level expiry risk.

    For each medicine, batches are consumed in FEFO order at the current
    daily consumption rate. A batch's projected units-at-risk are the units
    that will still be on the shelf on its expiry date:

        demand_before_expiry = daily_rate * days_until_expiry
        units_consumed       = clamp(demand_before_expiry - units in earlier batches, 0, qty)
        units_at_risk        = qty - units_consumed

    Risk level: EXPIRED (already expired), HIGH (>= 50% of the batch at
    risk, or any units at risk and expiry within the urgent threshold),
    MEDIUM (some units at risk), LOW (none projected).
    """
    thresholds = app_settings.expiry_thresholds(conn)
    usage = inventory.consumption(conn)
    rows = inventory.batches(conn, include_empty=False)

    by_medicine: dict[int, list[dict]] = {}
    for row in rows:
        by_medicine.setdefault(row["medicine_id"], []).append(row)

    results = []
    for medicine_id, batch_list in by_medicine.items():
        daily = usage.get(medicine_id, {}).get("average_daily", 0) or 0
        earlier_units = 0

        for batch in sorted(batch_list, key=lambda b: (b["expiry_date"], b["batch_id"])):
            qty = batch["quantity"]
            days = batch["days_until_expiry"]

            if days < 0:
                at_risk = qty
                level = "EXPIRED"
            else:
                demand = daily * (days + 1)  # usable through the expiry date
                consumed = min(qty, max(0.0, demand - earlier_units))
                at_risk = int(round(qty - consumed))
                share = at_risk / qty if qty else 0
                if at_risk == 0:
                    level = "LOW"
                elif share >= 0.5 or days <= thresholds.urgent_days:
                    level = "HIGH"
                else:
                    level = "MEDIUM"
                earlier_units += qty

            cost = batch["unit_cost"]
            value_at_risk = round(float(cost) * at_risk, 2) if cost is not None else None

            if include_normal or level != "LOW":
                results.append({
                    "batch_id": batch["batch_id"],
                    "medicine_id": medicine_id,
                    "medicine": batch["medicine"],
                    "strength": batch["strength"],
                    "dosage_form": batch["dosage_form"],
                    "batch_number": batch["batch_number"],
                    "location": batch["location"],
                    "quantity": qty,
                    "expiry_date": batch["expiry_date"],
                    "days_until_expiry": days,
                    "expiry_status": batch["status"],
                    "average_daily_consumption": daily,
                    "projected_units_at_risk": at_risk,
                    "unit_cost": cost,
                    "stock_value": batch["stock_value"],
                    "value_at_risk": value_at_risk,
                    "risk_level": level,
                })

    order = {"EXPIRED": 0, "HIGH": 1, "MEDIUM": 2, "LOW": 3}
    results.sort(key=lambda r: (order[r["risk_level"]], -(r["value_at_risk"] or 0), r["days_until_expiry"]))
    return results


# ------------------------------------------------------------
# Valuation
# ------------------------------------------------------------

def valuation(conn: psycopg.Connection) -> dict:
    """Stock value at cost, split by expiry status and by location. Batches
    with no recorded unit cost are counted separately (never guessed)."""
    rows = inventory.batches(conn, include_empty=False)

    def blank():
        return {"units": 0, "value": 0.0, "batches": 0, "units_without_cost": 0}

    total = blank()
    by_status = {status: blank() for status in expiry.STATUS_ORDER}
    by_location: dict[str, dict] = {}

    for row in rows:
        for bucket in (total, by_status[row["status"]], by_location.setdefault(row["location"], blank())):
            bucket["units"] += row["quantity"]
            bucket["batches"] += 1
            if row["unit_cost"] is None:
                bucket["units_without_cost"] += row["quantity"]
            else:
                bucket["value"] = round(bucket["value"] + float(row["stock_value"]), 2)

    return {
        "total": total,
        "usable_value": round(total["value"] - by_status["EXPIRED"]["value"], 2),
        "expired": by_status["EXPIRED"],
        "by_expiry_status": by_status,
        "by_location": [{"location": name, **values} for name, values in sorted(by_location.items())],
    }


# ------------------------------------------------------------
# Demand forecast
# ------------------------------------------------------------

def forecast(conn: psycopg.Connection, horizon_days: int = 30) -> list[dict]:
    """Forecast demand with simple exponential smoothing (alpha 0.5) over the
    last 12 weekly dispensing totals. Returns the forecast for the horizon
    and whether usable stock covers it. With little history the forecast is
    flagged as low confidence."""
    rows = conn.execute(
        """
        WITH weeks AS (
            SELECT generate_series(0, 11) AS weeks_ago
        )
        SELECT medicines.id AS medicine_id, weeks.weeks_ago,
               COALESCE(SUM(sm.quantity), 0)::int AS units
        FROM medicines
        CROSS JOIN weeks
        LEFT JOIN batches ON batches.medicine_id = medicines.id
        LEFT JOIN stock_movements sm
            ON sm.batch_id = batches.id
            AND sm.movement_type = 'DISPENSED'
            AND sm.movement_date >= CURRENT_TIMESTAMP - (weeks.weeks_ago + 1) * INTERVAL '7 days'
            AND sm.movement_date <  CURRENT_TIMESTAMP - weeks.weeks_ago * INTERVAL '7 days'
        GROUP BY medicines.id, weeks.weeks_ago
        ORDER BY medicines.id, weeks.weeks_ago DESC
        """
    ).fetchall()

    series: dict[int, list[int]] = {}
    for row in rows:
        series.setdefault(row["medicine_id"], []).append(row["units"])  # oldest -> newest

    stock = {s["medicine_id"]: s for s in inventory.medicine_stock(conn)}
    results = []
    for medicine_id, weekly in series.items():
        level = None
        for units in weekly:
            level = units if level is None else 0.5 * units + 0.5 * level
        weekly_rate = level or 0.0
        demand = round(weekly_rate / 7 * horizon_days, 1)
        active_weeks = sum(1 for w in weekly if w > 0)
        item = stock[medicine_id]
        results.append({
            "medicine_id": medicine_id,
            "medicine": item["medicine"],
            "strength": item["strength"],
            "dosage_form": item["dosage_form"],
            "weekly_history_oldest_first": weekly,
            "forecast_weekly_demand": round(weekly_rate, 1),
            "forecast_demand": demand,
            "horizon_days": horizon_days,
            "usable_stock": item["usable_stock"],
            "projected_stock_at_horizon": round(item["usable_stock"] - demand, 1),
            "covers_horizon": item["usable_stock"] >= demand,
            "confidence": "LOW" if active_weeks < 4 else ("MEDIUM" if active_weeks < 8 else "HIGH"),
            "method": "Exponential smoothing (alpha 0.5) of 12 weekly dispensing totals",
        })
    results.sort(key=lambda r: (r["covers_horizon"], -r["forecast_demand"]))
    return results


# ------------------------------------------------------------
# Turnover and purchasing patterns
# ------------------------------------------------------------

def turnover(conn: psycopg.Connection) -> list[dict]:
    """Stock turnover over 90 days. Average stock is approximated as the mean
    of current and 90-days-ago stock, reconstructed from the ledger."""
    rows = conn.execute(
        """
        SELECT medicines.id AS medicine_id, medicines.name AS medicine,
               medicines.strength, medicines.dosage_form,
               COALESCE(SUM(batches_total.qty), 0)::int AS current_stock
        FROM medicines
        LEFT JOIN (SELECT medicine_id, SUM(quantity) AS qty FROM batches GROUP BY medicine_id) AS batches_total
            ON batches_total.medicine_id = medicines.id
        GROUP BY medicines.id
        """
    ).fetchall()
    changes = conn.execute(
        """
        SELECT batches.medicine_id,
               COALESCE(SUM(CASE WHEN sm.movement_type IN ('RECEIVED','RETURNED','ADJUSTMENT')
                                 THEN sm.quantity ELSE -sm.quantity END), 0)::int AS net_change_90d,
               COALESCE(SUM(sm.quantity) FILTER (WHERE sm.movement_type = 'DISPENSED'), 0)::int AS dispensed_90d
        FROM stock_movements sm
        JOIN batches ON batches.id = sm.batch_id
        WHERE sm.movement_date >= CURRENT_TIMESTAMP - INTERVAL '90 days'
        GROUP BY batches.medicine_id
        """
    ).fetchall()
    change_map = {c["medicine_id"]: c for c in changes}

    results = []
    for row in rows:
        change = change_map.get(row["medicine_id"], {"net_change_90d": 0, "dispensed_90d": 0})
        start_stock = row["current_stock"] - change["net_change_90d"]
        average = (row["current_stock"] + start_stock) / 2
        ratio = round(change["dispensed_90d"] / average, 2) if average > 0 else None
        results.append({
            **row,
            "stock_90_days_ago": start_stock,
            "average_stock": round(average, 1),
            "dispensed_90d": change["dispensed_90d"],
            "turnover_90d": ratio,
            "annualised_turnover": round(ratio * 365 / 90, 2) if ratio is not None else None,
            "days_of_inventory": round(90 / ratio, 1) if ratio else None,
        })
    results.sort(key=lambda r: (r["turnover_90d"] is None, -(r["turnover_90d"] or 0)))
    return results


def purchasing_patterns(conn: psycopg.Connection) -> dict:
    by_supplier = conn.execute(
        """
        SELECT suppliers.id AS supplier_id, suppliers.name AS supplier,
               COUNT(DISTINCT po.id) AS orders,
               COALESCE(SUM(poi.quantity_ordered * poi.unit_cost), 0) AS ordered_value,
               COALESCE(SUM(poi.quantity_received * poi.unit_cost), 0) AS received_value,
               COALESCE(SUM(poi.quantity_ordered), 0)::int AS units_ordered,
               COALESCE(SUM(poi.quantity_received), 0)::int AS units_received
        FROM suppliers
        LEFT JOIN purchase_orders po ON po.supplier_id = suppliers.id AND po.status <> 'CANCELLED'
        LEFT JOIN purchase_order_items poi ON poi.purchase_order_id = po.id
        GROUP BY suppliers.id
        ORDER BY ordered_value DESC
        """
    ).fetchall()
    lead_times = conn.execute(
        """
        SELECT po.supplier_id,
               ROUND(AVG(first_receipt.received::date - po.order_date), 1) AS average_lead_time_days
        FROM purchase_orders po
        JOIN (
            SELECT poi.purchase_order_id, MIN(pr.received_date) AS received
            FROM purchase_receipts pr
            JOIN purchase_order_items poi ON poi.id = pr.purchase_order_item_id
            GROUP BY poi.purchase_order_id
        ) AS first_receipt ON first_receipt.purchase_order_id = po.id
        GROUP BY po.supplier_id
        """
    ).fetchall()
    lead = {row["supplier_id"]: row["average_lead_time_days"] for row in lead_times}
    monthly = conn.execute(
        """
        SELECT to_char(date_trunc('month', po.order_date), 'YYYY-MM') AS month,
               COUNT(DISTINCT po.id) AS orders,
               COALESCE(SUM(poi.quantity_ordered * poi.unit_cost), 0) AS ordered_value
        FROM purchase_orders po
        LEFT JOIN purchase_order_items poi ON poi.purchase_order_id = po.id
        WHERE po.status <> 'CANCELLED'
        GROUP BY 1 ORDER BY 1
        """
    ).fetchall()
    return {
        "by_supplier": [
            {**row,
             "fill_rate_percent": round(row["units_received"] / row["units_ordered"] * 100, 1)
             if row["units_ordered"] else None,
             "average_lead_time_days": lead.get(row["supplier_id"])}
            for row in by_supplier
        ],
        "by_month": monthly,
    }


# ------------------------------------------------------------
# Dashboard
# ------------------------------------------------------------

def dashboard(conn: psycopg.Connection) -> dict:
    thresholds = app_settings.expiry_thresholds(conn)
    stock = inventory.medicine_stock(conn)
    value = valuation(conn)
    risk = expiry_risk(conn, include_normal=False)
    reorder = reorder_recommendations(conn, only_needed=True)
    movement = consumption_summary(conn)

    status_counts = {s: value["by_expiry_status"][s] for s in expiry.STATUS_ORDER}

    recent = conn.execute(
        """
        SELECT sm.id, sm.batch_id, medicines.name AS medicine, medicines.strength,
               medicines.dosage_form, batches.batch_number, sm.movement_type, sm.quantity,
               sm.movement_date, sm.reason, users.full_name AS user_name
        FROM stock_movements sm
        JOIN batches ON batches.id = sm.batch_id
        JOIN medicines ON medicines.id = batches.medicine_id
        LEFT JOIN users ON users.id = sm.user_id
        ORDER BY sm.movement_date DESC, sm.id DESC
        LIMIT 8
        """
    ).fetchall()

    return {
        "as_of": date.today(),
        "thresholds": thresholds.as_params(),
        "total_medicines": len(stock),
        "medicines_in_stock": sum(1 for s in stock if s["usable_stock"] > 0),
        "total_units": value["total"]["units"],
        "usable_units": sum(s["usable_stock"] for s in stock),
        "stock_value": value["total"]["value"],
        "usable_stock_value": value["usable_value"],
        "units_without_cost": value["total"]["units_without_cost"],
        "expired": status_counts["EXPIRED"],
        "critical": status_counts["CRITICAL"],
        "urgent": status_counts["URGENT"],
        "approaching": status_counts["APPROACHING EXPIRY"],
        "low_stock_count": sum(1 for s in stock if s["stock_status"] == inventory.LOW_STOCK),
        "out_of_stock_count": sum(1 for s in stock if s["stock_status"] == inventory.OUT_OF_STOCK),
        "slow_moving_count": sum(1 for m in movement if m["movement_class"] in ("SLOW-MOVING", "NO MOVEMENT") and m["usable_stock"] > 0),
        "expiry_risk": {
            "batches": len(risk),
            "units_at_risk": sum(r["projected_units_at_risk"] for r in risk),
            "value_at_risk": round(sum(r["value_at_risk"] or 0 for r in risk), 2),
            "top": risk[:6],
        },
        "reorder": reorder[:6],
        "reorder_count": len(reorder),
        "stock_levels": stock,
        "slow_moving": [m for m in movement if m["movement_class"] != "ACTIVE" and m["usable_stock"] > 0][:6],
        "recent_movements": recent,
        "valuation_by_location": value["by_location"],
    }
